cpp/src/groupby/common/utils.hpp

```diff
@@ -30,10 +32,24 @@ inline std::vector<aggregation_result> extract_results(host_span<RequestType con
                                                        cudf::detail::result_cache& cache)
 {
   std::vector<aggregation_result> results(requests.size());
-
+  std::unordered_map<std::pair<column_view, std::reference_wrapper<aggregation const>>,
+                     column_view,
+                     cudf::detail::pair_column_aggregation_hash,
+                     cudf::detail::pair_column_aggregation_equal_to>
+    repeated_result;
   for (size_t i = 0; i < requests.size(); i++) {
     for (auto&& agg : requests[i].aggregations) {
-      results[i].results.emplace_back(cache.release_result(i, *agg));
+      if (cache.has_result(requests[i].values, *agg)) {
+        results[i].results.emplace_back(cache.release_result(requests[i].values, *agg));
+        repeated_result[{requests[i].values, *agg}] = results[i].results.back()->view();
+      } else {
+        auto it = repeated_result.find({requests[i].values, *agg});
+        if (it != repeated_result.end()) {
+          results[i].results.emplace_back(std::make_unique<column>(it->second));
+        } else {
+          CUDF_FAIL("Cannot extract result from the cache");
+        }
```

Comment:
I don't follow why this else clause is necessary. Since the result cache's `_cache` uses the same shallow comparators as `repeated_result` here and you're inserting values into `repeated_result` from the cache, when would you find something in `repeated_result` that wouldn't have been found by querying the cache?

Reply:
`_cache` holds values as `std::unique_ptr<column>`, 
`repeated_result` holds values as `column_view`. 

This case happens when same aggregation request with same column_view is present more than once in the given vector of requests. Hashing will prevent from computing again but the result should be returned more than once. Hence this `repeated_result` is required.

Reply:
Sorry, my question was unclear. I was referring to the outer else clause conditioned on `has_result`, not the inner else clause that is only reached when the search through `repeated_result` fails. However, I figured out the source of my confusion. I did not realize that the the behavior of the if clause was that it _removed_ items from the cache when constructing the results (I missed the call to `extract`), I just thought it would point to them. I guess since we are returning values to the user we need to hand over ownership and clear cached data, so in the case of duplicates this approach lets us return separate columns rather than two views into the same underlying column.